### DISCUSSION TITLE
Adding "consumed" information for ECU-C

### DISCUSSION THREAD

**Discussion by gillesvs** [ksheumaker/homeassistant-apsystems_ecur#91] [Q&A]
Hello,

First, thank you very much for this integration, it works fine from the start with an ECU-C I installed last week..

I am not a developper, but I would love to know how to add the "consumed" information which is not part of it.

I understant that this integration was build firstly for an ECU-R which does not give Consumed Data, only production.
But the ECU-C does and it is to bad to add additional hardware in my electricity panel when the info is already collected with my ECU-C.

Any insight on how to add this data to the integration?

Thanks

**Comment by HAEdwin** [comment]
Hi Gillesvs, 
If I'm not mistaken, the ECU-C is installed via the built-in web interface. The integration is based on the communication between the ECU and the ECUapp (phone app) and no consumed energy parameter is available in that app. Without having an ECU-C myself, it is impossible for me to figure out how that parameter could be read. What I normally would do is sniffing the data that goes between the ECU-C and the configuration interface. Sorry but I can't help with this. 

Incidentally, the use of this integration in combination with the ECU-C is not stable. After a while, the ECU-C looses connection and must then be restarted. This is more of a firmware issue than an integration issue.

Should it be the case that the integration in combination with the ECU-C continues to run for a longer period of time, please let us know (incl. firmware version). Of course, it can always be the case that APSystems has improved this to our advantage.

**Reply by gillesvs** [reply]
Thank you for these info. I will tell you if the integration works in the long run without any restart from the ECU-C (which is not so much of a burden if it is once in a while.
And yes, consumed info are from the web interface , not the App. But as the APsystems EMA API documentation was refering to a way to get consumation and import/export, I was hoping for an API call solution (which I do not master I am afraid...)
Gilles

**Comment by gillesvs** [comment]
Hello

A last bit of information.

The integration was unstable indeed. And this morning, the ecu-c itself was unable to connect to my lan through wifi. A firmware update at 3am it seems.

I connected the ecu using rj45. I also assign the same ip address just to have a look at the integration ... And your integration is working fine in rj45 mode, without any wifi.

**Reply by gillesvs** [reply]
Firmware: ECU-C-Z_C1.2.2

**Reply by HAEdwin** [reply]
I understand that the ethernet connection for the ECU-C also works, but unfortunately it also will eventually stop functioning. There is an http option from tv3 available but it is not yet integrated with this integration. Some scedule a reset of the ECU to get past the issue. I do think it's an ECU issue not a integration issue. At around 3 AM the ECU's perform some maintenance and interaction with the EMA (all data present checks), sometimes this includes a firmware update. The older ECU-R and ECU-B recover from maintenance easily.

**Comment by redgtux** [comment]
Hi

You can retrieve json informations from ECU-C web interface by sending a POST request .
I tried on my ECU-C and it seems to be working :
![image](https://user-images.githubusercontent.com/58104131/199363407-f9827244-6ae7-4d0d-995a-0746087b8727.png)


Regards

**Reply by lopezmathieu** [reply]
Hell @redgtux did you try it on HA ? Does it works ? Interested cause I use a ECU-C too. Mathieu

**Reply by neoback45** [reply]
yes interested too!!

**Reply by lopezmathieu** [reply]
Just try, but seems it doesn't work. 
`{"APS":"ANSWER","VERSION":"1.1","error":1}`

**Reply by gregoo31** [reply]
I managed to recover the power consumed thanks to my MaxPV solar router.
everything works under HA.

https://github.com/Jetblack31/MaxPV
https://forum-photovoltaique.fr/viewtopic.php?f=110&t=55244&hilit=maxpv

**Reply by lopezmathieu** [reply]
> I managed to recover the power consumed thanks to my MaxPV solar router. everything works under HA.
> 
> https://github.com/Jetblack31/MaxPV https://forum-photovoltaique.fr/viewtopic.php?f=110&t=55244&hilit=maxpv

You got the info from the router or from ECU ?

**Reply by gregoo31** [reply]
the router helps out for now

**Comment by Geekarchitecte** [comment]
Hello everyone,

I am also looking for an integration for the ECU-C which allows to have all that is consumption.
With this one, I manage to reassemble perfectly all that is production (ECU and inverters), nice job.
For the moment it works pretty well, I haven't noticed any crashes in my ECU at least since the 1.2.26 integration
For information, the firmware version of my ECU is V5.5.0
It would be so great if someone could bring up the consumption, import and injection info (sorry, I don't have any other skills than possibly testing...).

**Reply by HAEdwin** [reply]
Since the consumed energy is not part of the displayed info in the manager app, it would be a wild guess what the APS command will be to retreive this info. Could be something like APS1100280053[ECU-ID]END where 28 is the length of the commandstring and 53 the command as shown above. ECU-ID = 12 characters long. Because I do not own a ECU-C I'm not able to test if this might return a string from the ECU-C. With the NC command as shown in the readme you could try if the ECU-C returns a string. But do this at your own risk because it is known that some commands can factory reset the ECU. If you dare to try and post the result, we might get a step closer to solving this wish. According to the json output it would return a timestamp and three values for power A to C.

**Reply by mberglundmx** [reply]
I **do** see the produced / imported energy in the manager app - or do I missunderstand what you mean? 
I'm not sure how to sniff out the correct command to try though - your wild guess above did not result in any response at all with my ECU-C.

If I understand correctly, you use the same commands as the manager app uses when connected to the local AP of the device? I'm not sure how I can wireshark / sniff that traffic? Is there any other way I can see what command is used to fetch that data? I could of course just try random commands but am a bit worried I will screw something up :)

**Comment by Geekarchitecte** [comment]
In my case, the energy consumed is well integrated into the EMA App application (and it also shows up on the website).
For the rest, I will try to understand and why not see...

**Reply by tediroca** [reply]
I'm uinsg an ECU-C and so far didn't get any of the consumptions in the App, website or ECU-C webserver, which I guess it's because my brand new system is still in zero-injection mode. 

Once it's showing the data in any of the panels I'll try to dig a bit more into it too.

**Reply by tediroca** [reply]
My installation seem to have started to show the data for consumptions from the grid too:
![image](https://user-images.githubusercontent.com/1898632/217204091-d9ac1e39-7a1c-404c-adce-1890f9a3232a.png)

Any idea of how we could help including these into the integration?

**Comment by lamsof** [comment]
Hello All

In the downloaded package, we can see a print screen showing the consumed data while in the code, and HMI, I cannot find it.
Also, the code seems to be based on the local APSystem API not the one on the web. I have never entered my apsystemema login/password to retreive the data, just the local IP adress (which doesn't need any login).
![image](https://user-images.githubusercontent.com/127216895/223489260-30c54089-59dd-4f15-999e-7df6bd75faf4.png)

Anyone was able to retreive the consumed power ? My goal is to identify if I'am not consuming the produced energy and reinject the extra power into my EV car (Tesla)

**Reply by neoback45** [reply]
Hello,
With "plugin" jeedom i have a consumed power. you have find with integration HA?

**Reply by gregoo31** [reply]
I bought the apsystem application for jeedom, the energy consumed is well indicated (I no longer use jeedom because many paid plugins are not followed by the developers... in particular that of the rflink for my somfy roller shutters) I start quietly with Home assistant ;-)

Can we retrieve the API request information in the jeedom apsystem application?
I have a very very low level in programming!

**Reply by gregoo31** [reply]
Good morning,
someone figured out how to get power consumption information (import/export) from ECU-C.
If someone can explain the steps that would be great.
Or integrate it into the application?
Can we also recover the consumption (alone) of the house without solar production (- or +)

Sorry but my level does not allow me this integration...

[https://github.com/ksheumaker/homeassistant-apsystems_ecur/issues/157#issuecomment-1491523320]

**Reply by mberglundmx** [reply]
I managed to get the data in the HA with a REST sensor - however this causes more communication with the ECU, which feels like it causes more frequent crashes, and the plugin fails to restart it - which means I need to manually go out and reset it.

Due to this I have removed the rest sensor for now, would be better to get it as part of this integration.

**rest.yaml**
```
- resource: http://<ECU_IP>/index.php/meter/old_meter_power_graph
  scan_interval: 300
  sensor:
    - name: power_produced_aps_L1
      value_template: '{{ value_json["power1"][-1]["powerA"] }}'
      unit_of_measurement: W
      state_class: measurement
    - name: power_produced_aps_L2
      value_template: '{{ value_json["power1"][-1]["powerB"] }}'
      unit_of_measurement: W
      state_class: measurement
    - name: power_produced_aps_L3
      value_template: '{{ value_json["power1"][-1]["powerC"] }}'
      unit_of_measurement: W
      state_class: measurement
    - name: net_power_L1
      value_template: '{{ value_json["power2"][-1]["powerA"] }}'
      unit_of_measurement: W
      state_class: measurement
    - name: net_power_L2
      value_template: '{{ value_json["power2"][-1]["powerB"] }}'
      unit_of_measurement: W
      state_class: measurement
    - name: net_power_L3
      value_template: '{{ value_json["power2"][-1]["powerC"] }}'
      unit_of_measurement: W
      state_class: measurement
```

**configuration.yaml**
```
rest: !include rest.yaml
```

**Reply by gregoo31** [reply]
I tested your code rest.yaml
and everything seems to work.
I recover my net consumption well!
THANKS

**Reply by neoback45** [reply]
> I tested your code rest.yaml
> 
> and everything seems to work.
> 
> I recover my net consumption well!
> 
> THANKS

Good job! But it's stable ?
Not bug ?
